File: src/aside/java/org/taktik/icure/logic/EntityPersister.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import org.taktik.icure.validation.aspect.Check;

import java.util.Collection;
import java.util.List;

public interface EntityPersister<E, I> {

	boolean createEntities(@Check Collection<E> entities, Collection<E> createdEntities) throws Exception;

	List<E> updateEntities(@Check Collection<E> entities) throws Exception;

	void deleteEntities(Collection<I> identifiers) throws Exception;

	void undeleteEntities(Collection<I> identifiers) throws Exception;

	List<E> getAllEntities();
	List<String> getAllEntityIds();

	boolean hasEntities();

	boolean exists(String id);

	E getEntity(String id);
}

File: src/main/kotlin/org/taktik/icure/asynclogic/EntityPersister.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.DocIdentifier

interface EntityPersister<E, I> {

    fun createEntities(entities: Collection<E>): Flow<E>

    fun modifyEntities(entities: Collection<E>): Flow<E>

    fun deleteEntities(identifiers: Collection<I>): Flow<DocIdentifier>
    fun undeleteByIds(identifiers: Collection<I>): Flow<DocIdentifier>

    fun getEntities(): Flow<E>
    fun getEntityIds(): Flow<I>

    suspend fun hasEntities(): Boolean

    suspend fun exists(id: I): Boolean

    suspend fun getEntity(id: I): E?
}
